### DISCUSSION TITLE
What is the best way to programmatically know what module to set `FLASK_APP` to?

### DISCUSSION THREAD

**Discussion by brettcannon** [pallets/flask#4660] [Q&A] [answered]
For the [Python extension in VS Code](https://marketplace.visualstudio.com/items?itemName=ms-python.python), we want to provide automatic [debug configurations for Flask](https://code.visualstudio.com/docs/python/debugging#_flask-debugging) users when we can easily do so. That means (we think) needing to know what module to set `FLASK_APP` to. Is there a best practice for determining this or something we are missing that we can use to determine this? Or is there something else we can rely on to do the right thing where we can control how Python is executed?

**Comment by ThiefMaster** [comment]
Some ideas that come to my mind:

- Look for a module that contains `app = Flask(...)` (ie something named `app` with `isinstance(app, Flask)`)
- Look for a module that contains a `create_app()` or `init_app()` function that returns a Flask instance
- Look for a `.env` or `.flaskenv` that sets `FLASK_APP` and use that

**Reply by brettcannon** [reply]
> Some ideas that come to my mind

Thanks for the ideas!



> * (ie something named `app` with `isinstance(app, Flask)`)

We would like to avoid executing Python code since the debugger is launched via TypeScript and a created shell command. But we could run a regex over some files to look for `app\s*=\s*Flask\(`.

**Reply by ThiefMaster** [reply]
I was more thinking something in the typing/mypy direction to figure out if something is a Flask instance (or instance of a Flask subclass) ;)

**Comment by davidism** [comment, marked as answer]
Flask itself has some detection for when `FLASK_APP` isn't set that will probably serve you well enough.

1. Look for a file called `wsgi.py` or `app.py` in the current directory.
2. Within that, look for a variable called `app` or `application`, if it's a `Flask` object return it.
3. Look for any variable that is an instance of `Flask` and return that.
4. Look for a function called `create_app` or `make_app`. Call it without arguments, if it returns a `Flask` object return it.

I'd suggest extending this to looking in `__init__.py` and `app.py` files within the packages in the current project, something VS Code should be able to figure out which Flask itself can't. And since you're not actually executing code, look for `var = Flask(` like you suggested, and assume factory functions return correctly.

If you can't detect something automatically using the rules above, or even if you can, I'd suggest looking at PyCharm Professional's Flask server run configuration. It can detect what to run in many cases (probably the rules above), and then allows customizing them further to support the full range of what `FLASK_APP` and `flask run` can do.

**Reply by brettcannon** [reply]
What we are thinking of doing is searching for `^app(?:lication)?\s*=\s*=(?:flask\.)Flask\(` or `def\s+(?:create|make)_app\(` in `wsgi.py`, `app.py`, `*/__init__.py`, `*/app.py`, `src/*/__init__.py`, and `src/*/app.py`.

**Reply by davidism** [reply]
@brettcannon not sure if this is handled by VS Code, but it's related to running the development server. I'm deprecating the `FLASK_ENV=development` environment variable and going back to using `FLASK_DEBUG=1` (or the `flask --debug run` option as well now) to control debug mode. If you're generating a run configuration, you might want to enable debug mode by default.

If you're running with the VS Code debugger, you might want to disable the Flask debugger and reloader as well, using the `flask --debug run --no-debugger --no-reload` options. All options can also be controlled with environment variables, like `FLASK_RUN_DEBUGGER=0` and `FLASK_RUN_RELOAD=0`.

**Comment by paulacamargo25** [comment]
Thanks for you reply @davidism, I was going through the arguments we have in the python extension debugger and wanted to ask why in the previous comment you suggested disabling the Flask debugger and reloader. Since on the user side it is better if these are activated.  So I wanted to confirm if there is a known bug for leaving these options active, so I can leave them active by default.

Thanks

**Reply by davidism** [reply]
If you're using the IDE's debugger, you don't also need the built-in debugger UI. If you're using breakpoints, you probably don't want the server to restart in the middle of a debugging session. Both can remain enabled, they just also make sense disabled in this case.
